homeassistant/helpers/network.py

```diff
@@ -1,38 +1,230 @@
+    """An URL to the Home Assistant instance is not available."""
+
+
+@bind_hass
+@callback
+def async_get_url(
+    hass: HomeAssistant,
+    *,
+    require_ssl: bool = False,
+    require_standard_port: bool = False,
+    allow_internal: bool = True,
+    allow_external: bool = True,
+    allow_cloud: bool = True,
+    allow_ip: bool = True,
+    prefer_external: bool = False,
+    prefer_cloud: bool = False,
+) -> str:
+    """Get a URL to this instance."""
+    order = [TYPE_URL_INTERNAL, TYPE_URL_EXTERNAL]
+    if prefer_external:
+        order.reverse()
+
+    # Try finding an URL in the order specified
+    for url_type in order:
+
+        if allow_internal and url_type == TYPE_URL_INTERNAL:
+            try:
+                return _async_get_internal_url(
+                    hass,
+                    allow_ip=allow_ip,
+                    require_ssl=require_ssl,
+                    require_standard_port=require_standard_port,
+                )
+            except NoURLAvailableError:
+                pass
+
+        if allow_external and url_type == TYPE_URL_EXTERNAL:
+            try:
+                return _async_get_external_url(
+                    hass,
+                    allow_cloud=allow_cloud,
+                    allow_ip=allow_ip,
+                    prefer_cloud=prefer_cloud,
+                    require_ssl=require_ssl,
+                    require_standard_port=require_standard_port,
+                )
+            except NoURLAvailableError:
+                pass
+
+    # We have to be honest now, we have no viable option available
+    raise NoURLAvailableError
 
 
 @bind_hass
 @callback
-def async_get_external_url(hass: HomeAssistant) -> Optional[str]:
-    """Get external url of this instance.
+def _async_get_internal_url(
+    hass: HomeAssistant,
+    *,
+    allow_ip: bool = True,
+    require_ssl: bool = False,
+    require_standard_port: bool = False,
+) -> str:
+    """Get internal URL of this instance."""
+    if hass.config.internal_url:
+        internal_url = yarl.URL(hass.config.internal_url)
+        if (
+            (not require_ssl or internal_url.scheme == "https")
+            and (not require_standard_port or internal_url.is_default_port())
+            and (allow_ip or not is_ip_address(internal_url.host))
+        ):
+            return normalize_url(str(internal_url))
+
+    # Fallback to old base_url
+    try:
+        return _async_get_deprecated_base_url(
+            hass,
+            internal=True,
+            allow_ip=allow_ip,
+            require_ssl=require_ssl,
+            require_standard_port=require_standard_port,
+        )
+    except NoURLAvailableError:
+        pass
+
+    # Fallback to detected local IP
+    if allow_ip and not (
+        require_ssl or hass.config.api is None or hass.config.api.use_ssl
+    ):
+        ip_url = yarl.URL.build(
+            scheme="http", host=get_local_ip(), port=hass.config.api.port
```

Comment:
`get_local_ip` does I/O https://github.com/home-assistant/core/blob/1d6a20228be421c140d3afbaf33a00d419bec5e3/homeassistant/util/__init__.py#L95-L101

Reply:
😓

Reply:
We should cache it up front, maybe during startup?

Reply:
Looking at the codebase, this is doing blocking I/O in quite a few places.

Affected components:

- `dlna_dmr`
- `emulated_hue`
- `emulated_roku`
- `homekit` (potentially)
- `http`

Suggestion: Add it to the ApiConfig on HTTP start (config.api.local_ip)

Reply:
Done. Made a task to resolve other occurrences in the codebase in follow up PR.